integrations/pgvector/src/haystack_integrations/document_stores/pgvector/filters.py

```diff
@@ -0,0 +1,228 @@
+# SPDX-FileCopyrightText: 2023-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+from datetime import datetime
+from itertools import chain
+from typing import Any, Dict, List
+
+from haystack.errors import FilterError
+from pandas import DataFrame
+from psycopg.sql import SQL
+from psycopg.types.json import Jsonb
+
+# we need this mapping to cast meta values to the correct type,
+# since they are stored in the JSONB field as strings.
+# this dict can be extended if needed
+PYTHON_TYPES_TO_PG_TYPES = {
+    int: "integer",
+    float: "real",
+    bool: "boolean",
+}
+
+
+def _convert_filters_to_where_clause_and_params(filters: Dict[str, Any]) -> tuple[SQL, tuple]:
+    """
+    Convert Haystack filters to a WHERE clause and a tuple of params to query PostgreSQL.
+    """
+    if not isinstance(filters, dict):
+        msg = "Filters must be a dictionary"
+        raise FilterError(msg)
+
+    if "field" in filters:
+        query, values = _parse_comparison_condition(filters)
+    else:
+        query, values = _parse_logical_condition(filters)
+
+    where_clause = SQL(" WHERE ") + SQL(query)
+    params = tuple(value for value in values if value != "no_value")
+
+    return where_clause, params
+
+
+def _parse_logical_condition(condition: Dict[str, Any]) -> tuple[str, List[Any]]:
+    if "operator" not in condition:
+        msg = f"'operator' key missing in {condition}"
+        raise FilterError(msg)
+    if "conditions" not in condition:
+        msg = f"'conditions' key missing in {condition}"
+        raise FilterError(msg)
+
+    conditions = [_parse_comparison_condition(c) for c in condition["conditions"]]
+    query_parts, values = [], []
+    for c in conditions:
+        query_parts.append(c[0])
+        values.append(c[1])
+    if isinstance(values[0], list):
+        values = list(chain.from_iterable(values))
+
+    operator = condition["operator"]
+    if operator == "AND":
+        sql_query = "(" + " AND ".join(query_parts) + ")"
+    elif operator == "OR":
+        sql_query = "(" + " OR ".join(query_parts) + ")"
+    else:
+        msg = f"Unknown logical operator '{operator}'"
+        raise FilterError(msg)
+
+    return sql_query, values
+
+
+def _parse_comparison_condition(condition: Dict[str, Any]) -> tuple[str, List[Any]]:
+    if "field" not in condition:
+        # 'field' key is only found in comparison dictionaries.
+        # We assume this is a logic dictionary since it's not present.
+        return _parse_logical_condition(condition)
```

Comment:
This is confusing to me: something called `_parse_comparison_condition` might return the parsing of a logical condition... Shouldn't be the caller making this check?

Reply:
you are right...
(copy-pasted from another Document Store, but the caller already checks this)

Reply:
agree!